lib/prometheus/middleware/collector.rb

```diff
@@ -50,15 +52,15 @@ def call(env) # :nodoc:
       COUNTER_LB = proc do |env, code|
         {
           code:   code,
-          method: env['REQUEST_METHOD'].downcase.scrub('_NON_UTF8_'),
-          path:   aggregation.call(env['PATH_INFO']).scrub('_NON_UTF8_'),
+          method: env['REQUEST_METHOD'].downcase.gsub(/[^#{Regexp.escape RFC3986_CHARACTERS}]/, '__NON_RFC3986_CHAR__'),
```

Comment:
I wonder if we should not be even stricter than this.

For instance, this fails:
> "\xFF".gsub(/[^#{Regexp.escape RFC3986_CHARACTERS}]/, '\_\_NON_RFC3986_CHAR\_\_')
ArgumentError: invalid byte sequence in UTF-8

By the way, even this fails:
> "\xFF".downcase
ArgumentError: input string invalid

I suggest the following to be super safe:
> env['REQUEST_METHOD'].encode("ASCII", "UTF-8", invalid: :replace, undef: :replace, replace: "\_\_NON_RFC3986_CHAR\_\_").downcase.gsub(/[^#{Regexp.escape RFC3986_CHARACTERS}]/, '\_\_NON_RFC3986_CHAR\_\_')

It works with "\u001C" and "\xFF".
WDYT?

Reply:
It is because \xFF is invalid whereas u001C is valid... using #scrub in addition would be sufficient (I tried it), but you're right, we should re-encode to ASCII as all allowed chars are ASCII anyway!

Btw I wanted to add tests but the test stack and URI module do not allow all those invalid sequences, so I'm not even sure how it manages to arrives here in production...